### ISSUE TITLE
Improving The Platform Extension System

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#3515]
I'd like to start discussing how we can further improve the system by which different platforms are added in MonoGame.

IMO the idealistic goal moving forward should be that we can add additional platforms (like PS4 for instance) without needing to make platform specific modifications to existing files in the public repo.  Ideally merging in a platform like XB1, PS4, or WiiU would be as simple as copying a folder of files into the root of the public repo and executing Protobuild.

About a year ago we took two big steps...
## Moved to using Protobuild for project generation

The Protobuild change now means zero project file maintenance, new non-platform specific code is quick to add, and no more project file conflicts.  

The problem with Protobuild is that all the platform specific info is centralized in one file.  This means that if we have a closed source platform like we do with PS4... we have to maintain a patched version of that definition file with additional `.ps4.cs` files and other platform specific details.

Ideally there should be a way to optionally include additional platform files/settings from the main Protobuild .definition file.  @hach-que ??
## Partial classes and "Platform" methods

Also a lot of MonoGame platform code is now done with partial classes and "Platform" methods.  This is great as most of the platform code is just adding in various `.platform.cs` and editing the protobuild config file to include them.

We still have big classes like `GraphicsDeviceManager` that are a `#if` spaghetti.  We should prioritize converting classes like that into platform partial classes.

In places where we smaller `#if` blocks that are enabling some feature for multiple platforms we should look at add a feature specific `#if`.  For example this case here:

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Game.cs#L325

... we probably need a `#if NO_EXIT` instead of naming specific platforms.  This allows the feature to be toggled on other platforms by changing the Protobuild config.
## Extending The Pipeline

A big area where we need to design a new abstraction is within the content pipeline.  There are cases where there are platform specific formats and processing that we want to be able to plug into the pipeline instead of being hard coded into the public repo.

In this case I started a separate issue for this (see https://github.com/mono/MonoGame/issues/2781) but it is something that needs consideration.

**Comment by hach-que** [comment]
Yeah, we can update Protobuild to use a file like `Build\GenerateProject.CSharp.<Platform>.definition` for a specific platform if it exists (this is something I've been thinking about for a while on how to support NDA'd platforms).

**Comment by tomspilman** [comment]
> Protobuild to use a file like `Build\GenerateProject.CSharp.<Platform>.definition` for a specific platform if it exists

Cool... sounds like we've been thinking of similar things. :)

However it would be good if the `.<Platform>.definition` file could somehow be in a separate folder.  This way for instance you can have a private repo with files for some NDA'd platform and it can be added as a child folder and/or submodule of the MonoGame repo and it be automatically picked up by Protobuild.

No clue how this would work, but I think it is critical to making closed platforms modular additions.

**Comment by hach-que** [comment]
Do you not continuously merge the open source master branch into the PS4 fork?  You shouldn't have conflicts in the Build folder because only the PS4 fork will have that file.

**Comment by tomspilman** [comment]
> Do you not continuously merge the open source master branch into the PS4 fork?

We do... but that breaks down beyond one NDA'd platform.

Think of the case where you are building a game and have multiple NDA'd platforms to support.  Do we require every developer to manually merge multiple NDA'd repos into their own to have a single repo for their project?

This is why i'm thinking for NDA'd platforms we want everything within a single subfolder folder.  It is extremely easy to continuously maintain multiple platforms together in single code base.

**Comment by hach-que** [comment]
I'm assuming for the NDA'd platforms you still want a single runtime DLL?

**Comment by tomspilman** [comment]
> I'm assuming for the NDA'd platforms you still want a single runtime DLL? 

Yes.

**Comment by hach-que** [comment]
This is my strategy so far; on the Protobuild side:

 * Allow custom generation XSLT per platform by using GeneratePlatform.CSharp.<platform>.xslt if it exists.
 * Add a “ProjectExtension” package type into Protobuild.  Packages downloaded with this type are flagged locally so that Protobuild will look inside the project extensions’ Build folder for additional generation XMLs (GenerateProject.CSharp.<platform>.xslt)
 * Provide a “Closed Source” option for Protobuild packages which indicates that a failure to clone is acceptable (because the user may not have access).
 * Support a ProjectExtension definition with AppliesTo attribute, which looks like:

```xml
<?xml version="1.0" encoding="utf-8"?>
<ProjectExtension AppliesTo="MonoGame.Framework" Path="MonoGamePS4">
  <Properties>
    <!-- A list of properties to add -->
  </Properties>
  <References>
    <!-- A list of references to add -->
  </References>
  <Files>
    <!-- A list of file entries to add, e.g. -->
    <Compile Include="SomePS4Code.cs">
      <Platforms>PS4</Platforms>
    </Compile>
  </Files>
</Project>
```

 * Files tags that are generated in the project are flagged with <FromProjectExtension /> and the full path to the project extension.  We then ignore entries with those tags when synchronising back (so that files coming from a project extension never go back into the original project).

On the MonoGame side:

 * Create a private repo with the additional files and project extensions for the NDA’d platform.
 * Create a package on http://protobuild.org/ that points to this repo, and flag the package as “Closed Source”.
 * Add the NDA’d project extension to open source MonoGame with “Protobuild.exe --add http://protobuild.org/MonoGame/<package name>”.
 * Users who do have access to the private repo will get the package source code, which will in turn provide the code / project generation for that platform.
 * Users who do not have access to the private repo won’t be affected because Protobuild will ignore clone failures when trying to clone a closed source package.

Does this sound reasonable?

**Comment by tomspilman** [comment]
> Create a package on http://protobuild.org/ that points to this repo

I really don't like the idea of needing the support of some other website to fetch code like this.  I know this is your new thing and i'm sure it will be great for most...  but it won't be something i'll ever get into.

This needs to work all locally if the files exist in the right place.

**Comment by hach-que** [comment]
You can set up your own package index at http://pkg.monogame.net or
something like that if you want.

Users can always clone the repo into the folder as well and it would work,
but you can't use Git submodules because they won't handle failure properly
(not to mention you'd have to update what the latest commit is in the
MonoGame repo whenever you update the code).

**Comment by tomspilman** [comment]
> Users can always clone the repo into the folder as well and it would work,

As long as that works and yet another service/server isn't required... great.

So where would this `<ProjectExtension>` file be located?  How does Protobuild.exe in the root of the MonoGame repo find it?

**Comment by hach-que** [comment]
In the Build\Projects folder of the PS4 Git repository.  So you'd have Build\Projects\MonoGamePS4.definition and that would contain the project extension information.

**Comment by tomspilman** [comment]
> In the Build\Projects folder of the PS4 Git repository

So assume we had the following structure:

 - MonoGame_Public_repo
  - Build
  - MonoGame_PS4
    - Build
 - Protobuild.exe

I guess running Protobuild.exe in the root of the MonoGame folder will automatically find the nested second `MonoGame_PS4/Build` folder?

Does Protobuild.exe just search recursively for `Build` folders?

**Comment by hach-que** [comment]
Protobuild would reside next to both build folders, basically you have Protobuild in the root of all modules / repositories.

The MonoGame public repo is actually like this:

 * MonoGame public repo
   * Build
   * Protobuild.exe

With PS4 it would be like:

* MonoGame public repo
  * Build
  * MonoGame PS4 repo
    * Build
    * Protobuild.exe
  * Protobuild.exe

**Comment by tomspilman** [comment]
So you would need this extra `Protobuild.exe` even if it isn't used?  Since the PS4 repo only has a `<ProjectExtension>` it would require the parent repo for it to work.

**Comment by hach-que** [comment]
Yes, Protobuild modules are defined by:

  * The presence of Protobuild.exe
  * The presence of Build\Module.xml

The reason it works like this is that you could take Protobuild.exe, modify it in almost any way you want (with a few exceptions), place it in the MonoGame repository, and then for anyone else uses Protobuild who has MonoGame as a submodule, it would still work perfectly.  Thus this allows absolute flexibility; if you can't customize something through modified XSLT and you need to change Protobuild's C# behaviour, you can.

In this case, Protobuild would be invoked in the submodule like it would for any other.  You could even define projects in the PS4 repos to provide some custom or closed source PS4 utilities, and they would be included in the main MonoGame solution after generation for use.

**Comment by tomspilman** [comment]
> The reason it works like this is that you could take Protobuild.exe, modify it in almost any way you want 

I see... the reason is that each Protobuild.exe will execute any nested Protobuild.exe it finds since it could be customized to process differently.

Well nothing is moving on this just yet, but in the coming months we'll want to look to supporting this `<ProjectExtension>` system like you've defined.  It is a key part of supporting more platforms that I think is pretty critical.

Like other engines still just dump a whole new copy of game code hacked with some console platform.  It is up to you to merge multiple platforms together and keep stuff up to date.

Working on a clean way to do this in MonoGame just reduces the pain points for developers.

**Comment by hach-que** [comment]
Just a heads up, the latest version of Protobuild now supports platform XSLT files.

Basically, just have a file named "GenerateProject.CSharp.Playstation4.xslt" and it will be used for the Playstation4 platform instead of "GenerateProject.CSharp.xslt" (or "GenerateProject.xslt").  You can use whatever platform names you want.

**Comment by tomspilman** [comment]
@hach-que - This issue is becoming critical moving forward into next year.

Did you ever look into implementing the `<ProjectExtension>` feature you described here:

  https://github.com/mono/MonoGame/issues/3515#issuecomment-73843280

**Comment by hach-que** [comment]
I haven't had a chance to make it work with `<ProjectExtension>` yet.  I'm guessing `GenerateProject.CSharp.Playstation4.xslt` isn't a suitable solution in the short term because of NDAs?

**Comment by tomspilman** [comment]
>  I'm guessing  GenerateProject.CSharp.Playstation4.xslt  isn't a suitable solution in the short term because of NDAs?

I guess I don't understand how that solves the issue.

No PS4 code, projects, or even names of APIs or file names of any kind can be in the public MonoGame repo.  How do I use `GenerateProject.CSharp.Playstation4.xslt` to achieve that?

**Comment by hach-que** [comment]
Yeah, that's what I mean; NDAs prevent `GenerateProject.CSharp.Playstation4.xslt` from being in the public repository (which is how it currently works without `ProjectExtension`).

**Comment by tomspilman** [comment]
> Yeah, that's what I mean; NDAs prevent  GenerateProject.CSharp.Playstation4.xslt  from being in the public repository

Yes.

The most that could be in the public repo would be that the PS4 project exists and what the expected location of the folder with `xslt`s, `<ProjectExtension>` files, code, etc. lives.

**Comment by hach-que** [comment]
This isn't on the top of the priority list, but only because I don't have access to any NDA'd platforms (so it makes it extremely difficult to architecture it correctly).

I'll add a Protobuild issue for this (since it's not being tracked at the moment).  Project extensions are almost identical to [Include Projects](https://protobuild.readthedocs.org/en/latest/include_projects.html), except that Include Projects don't support adding references or properties, and they require the project to reference the include project (rather than the project extension which has AppliesTo).

**Comment by hach-que** [comment]
(I'm open to receiving PRs to implement it though)

**Comment by tomspilman** [comment]
> (I'm open to receiving PRs to implement it though)

 I'll see if I can find some time next month for that.  But most likely i'll implement something you won't be happy with if you leave me to my own devices...  we really should work out the design in https://github.com/hach-que/Protobuild/issues/133 to make this productive.

**Comment by hach-que** [comment]
The relevant issues are:

https://github.com/hach-que/Protobuild/issues/133
https://github.com/hach-que/Protobuild/issues/134
https://github.com/hach-que/Protobuild/issues/135
https://github.com/hach-que/Protobuild/issues/136

We should address them individually; I think https://github.com/hach-que/Protobuild/issues/134 is probably the most practical to get started with.

**Comment by DDReaper** [comment]
For protobuild testing, you shouldn't need an actual NDA platform to test with, so long as the <ProjectExtension> feature enables platform extension or platform replacement as a feature.

Basically so I can have one project, then have an extension project which builds a replica but with new platform specifics.

**Comment by hach-que** [comment]
The problem we run into is when you have multiple projects that want to
support PS4, of which MonoGame is only one. For example if I had a unit
testing framework and I wanted it to target PS4.

If multiple projects include a PS4 platform template, then it's ambiguous
as to which one to pick. They may be different, this issue is amplified by
the fact that everything is private, so projects may not even know that
each team is supporting PS4.

We can't solve this issue by making MonoGame use the (in this example)
testing framework, and we can't solve the issue by making the testing
framework depend on MonoGame (this would add the dependency for all
platforms, and in general the dependency doesn't make sense).

The solution to this problem is to put the platform template in its own
module, and have each project depend on that module. Protobuild will
recognize that the reference is the same and unify it so there's only one
copy (we're assuming there's a top level module here that depends on both
MonoGame and the unit testing framework).

However now we need to address the issue of authorizing the user to
download the platform template module. How do we know the user is allowed
(has signed the NDA) to download or git clone the platform template?
Remember that we can't devise a solution that requires manually authorizing
users because it needs to be practical for all projects.

The solution to that problem is outlined in
https://github.com/hach-que/Protobuild/issues/135. In this case we're not
disclosing anything about the SDK, it's structure or how it works. Instead
we take a file installed on the system (that only people with the SDK will
have), and use the hash of that file as the seed for generating a
private/public key pair. The private key remains on the user's computer (in
memory), and is used to sign requests that ask to download the module with
the platform template in it. This same public private key pair can be used
for ssh to access the repository for people that want to git clone the
template.

**Comment by hach-que** [comment]
(I should point out that I'm talking about the GenerateProject.PlayStation4.CSharp.xslt file here, not the `<ProjectExtension>` semantic)

**Comment by tomspilman** [comment]
@hach-que - It seems simple to me.  We add something like this into Module.xml in the public MonoGame repo:

```
<ProjectExtensionPath>..\MonoGame.PS4</ProjectExtensionPath>
<ProjectExtensionPath>..\MonoGame.PSVita</ProjectExtensionPath>
```

Protobuild when it is run looks to see if those folders exist and if so look for a Build folder there which contains the "ProjectExtension" definition as well as necessary XSLTs and additional references and definitions specific to that platform.

None of this requires downloads, private/public keys, etc.   All we need is the project extension feature.

**Comment by hach-que** [comment]
The problem is say that I have a game that's built on MonoGame and I have the following dependencies:

* MonoGame (this repo)
* Jitter (a physics library)

When the solution is generated, we need to generate the Jitter library using the PS4 template.  The problem is that Jitter doesn't know about MonoGame because it's across from it in the hierarchy, so it won't be able to load the XSLT and thus the project won't generate correctly in the solution.

If we said Jitter could work by also depending on MonoGame.PS4 (i.e. you have one copy under MonoGame and one copy under Jitter), then you're still going to run into the problem that in order for people to use Jitter without MonoGame on PS4, they'd need to have access to the MonoGame PS4 repository.

It's important that we come up with a solution that will work for all projects that a game depends on, not just the MonoGame dependency itself.  Otherwise we'll have a situation where MonoGame might be portable to PS4, but developers will have to manually port and manage all of the other libraries that they're using (even where they're technically compatible).

**Comment by KonajuGames** [comment]
Wouldn't Jitter (in this example) have its own ProjectExtension for PS4 to
generate the PS4 version?  Or do I just not understand how this all works?​

**Comment by hach-que** [comment]
The problem isn't the project extension (this is almost already solved with include projects).  The issue is the XSLT to generate the PS4 C# project itself.  That's what needs to be shared / available to all projects in the solution.

**Comment by tomspilman** [comment]
> That's what needs to be shared / available to all projects in the solution.

It should assume that is provided by the PS4 repo in it's Build folder.  That seems simple and obvious.

**Comment by hach-que** [comment]
Which PS4 repository?  The MonoGame.PS4 repository?

Remember that the hierarchy that we're talking about looks like this:

* Game
  * MonoGame
    * MonoGame.PS4
  * Jitter

When Protobuild runs in the Jitter folder, it doesn't know anything about Game, MonoGame or MonoGame.PS4.  These folders are all above it (or above then down / across from it) in the filesystem hierarchy.  We don't search up the filesystem hierarchy in Protobuild beyond the module root (which is Jitter when it's running in Jitter), because this effectively means searching the entire machine for dependencies (Game will be under Documents will be under Users will be under C:, then we'd need to search down every folder from there just like we would need to search down from Game to MonoGame to MonoGame.PS4 to find the module that provides the platform XSLT).

This means in order for Jitter to know about MonoGame.PS4, it would need to depend on it in some way - whether that's as a git submodule (breaking non-PS4 platforms), a package reference, or getting users to copy MonoGame.PS4 into every third-party library they use (not something I want to support / recommend because it's tedious, error-prone, single-use / highly-specific and leaves the Git working directory unclean).

**Comment by hach-que** [comment]
The alternative here is to have the XSLT by itself in it's own repository; this is the recommendation that I  made previously - but we need to know the user is authorized to download this repository in that case, and I don't want the authorization list to be something that needs to be manually maintained by MonoGame (because then you'll need to be handing authorization for everyone, not just those that have purchased MonoGame for PS4).

**Comment by tomspilman** [comment]
> because then you'll need to be handing authorization for everyone, not just those that have purchased MonoGame for PS4

Why is this our problem?  We (meaning the MonoGame project) have no need to support people not using MonoGame for PS4.

So from my point of view... there is no one to authorize beyond people authorized to use MonoGame for PS4.

Also note MonoGame for PS4 is free to all PlayStation developers.

**Comment by hach-que** [comment]
That's my point - it shouldn't be your problem or responsible to be authorizing users to access the PS4 template.  That is what I want to avoid.

In order to do that, we need to have some way of verifying that someone has the SDK installed before allowing them to pull from a (project inspecific) repository with the PS4 platform XSLT inside it.

**Comment by tomspilman** [comment]
>  it shouldn't be your problem or responsible to be authorizing users to access the PS4 template.

I suppose you are still looking at this from the "use Protobuild to generate your entire game project" view.  I only look at Protobuild from one perspective...  generating projects for MonoGame.

No one will have access to any of the PS4 bits unless we give them access.  There is no extra responsibility for us.  If you want access to the MonoGame for PS4 bits... and you are an authorized PS4 developers...  we give you access.

I don't need Protobuild to solve any of these issues for me... these things are solved.  Verifying access to Sony SDKs, cloning repos, or supporting anything beyond MonoGame itself 100% unnecessary to the feature we need here.

**Comment by hach-que** [comment]
I always look at the solution in terms of using Protobuild to generate your entire game project - that is it's primary use-case for me.  I don't want to come up with some solution that makes MonoGame work, but leaves game developers with the same project management problems MonoGame used to have before switching to Protobuild.  And right now, that's a scenario that a lot of developers face if they're using MonoGame through the installer or NuGet - they have one project per platform, and manage it all manually.  We both know that this kind of project management is tedious and a waste of developers time.

In the context of supporting projects generated with Protobuild, we need a solution that will work for the other third party libraries that game developers are using - things like physics libraries, JSON serializers, etc. these libraries don't depend on MonoGame directly; either for desktop or console platforms.  Largely these libraries will work with a pretty general subset of the .NET APIs, but unless the projects are generated for the PS4 platform, they won't compile or link properly.

All we need here is to put the PS4 XSLT template in it's own private repository, and to provide authorised PS4 developers with a private key that can be used to access it - whether that's a private key generated from the SDK installation, distributed on some site that only PS4 developers can access (I don't know if there are forums or anything for this) - it doesn't matter.  But I don't want to saddle you guys with the responsibility of distributing it to authorized developers, because that's a waste of your time.

If I'm an authorized PS4 developer (and have the SDK installed), getting a game running on the PS4 with MonoGame should be as easy as:

```
Protobuild.exe --start http://protobuild.org/MonoGame/Template.2DPlatformer
Protobuild.exe --generate PlayStation4
```

Any more work than this and we're wasting the developer's time doing things that can and should be automated.

**Comment by KonajuGames** [comment]
There is no facility provided by the platform owners (Sony, Microsoft, etc)
that provides authentication for authorised developers. There is no
authenticated central file delivery system. Forums are usually custom, if
they exist, and change quite often. All vendors of libraries for these
closed platforms handle their own authentication, usually after a one time
manual check with the platform owner before providing a username/password,
before allowing licensees to download the library.

For other third party libraries, they most likely aren't using Protobuild.
If the developer wanted to use Protobuild with them, they round have to
generate a definition file for that library.

Restricting Protobuild to only look from the executable down is the
restricting factor here. How about similar to git or Perforce or other
source control systems where they will look up the directory tree for a
particular file or folder.  The extension could then be in the solution
root and all Protobuild instances down from there will be able to find it.

**Comment by hach-que** [comment]
Yes, that is why I believe using an SDK file as the basis for a private
public key pair is the most practical solution here; it doesn't require
central authentication.

There are a bunch of libraries like JSON.net and Jitter which are ported to
Protobuild, because this allows them to be used on all the platforms
Protobuild supports. This is one of the features I enjoy about Protobuild,
once you migrate the project system over, suddenly these libraries support
all platforms (and in the majority of cases work without requiring changes).

We can't only search immediately up, because this will break redirects
(they will search up and go "I'm not located in another module" even though
the other module references them outside the tree).

**Comment by KonajuGames** [comment]
Platform SDKs change a lot and frequently.  I wouldn't trust any file in a
platform SDK to be a good basis for a key.

**Comment by hach-que** [comment]
As far as I understand it, most SDKs sign their applications with a private
key so they can be deployed to the development kit. Since this key is
extremely unlikely to change (the public key is embedded into the dev kit
firmware), my plan was to use that key to digitally sign some string like
"PlayStation4" as verification that the user is capable of digitally
signing for the device (and is thus authorized).

**Comment by hach-que** [comment]
(also, how do you prove that someone has access to the SDK to give them
access to the MonoGame repository?)

**Comment by KonajuGames** [comment]
It is not the SDK that is signed but the executable that is produced. And
the signing may only be for the final master.

**Comment by KonajuGames** [comment]
And we don't have access to the signing key.

**Comment by hach-que** [comment]
If all else fails (we can't find a way of automatically verifying that users have SDK access without disclosing the SDK, and we can't manually verify without a central authority), then my fallback suggestion is to designate some folder in %appdata% as a location for additional platform XSLT files.  That folder would be searched, and people would just put `GenerateProject.PlayStation.CSharp.xslt` in there (so that it's available to all projects).  It's way less than ideal, but it might be the only option...

**Comment by hach-que** [comment]
Alright there's a new branch here: https://github.com/hach-que/Protobuild/tree/nda

This supports an AppliesTo attribute on include projects, which basically implements project extensions.  This doesn't support adding references yet, but I'm not sure how or what kinds of references you need to add (potentially we could add AppliesTo to external projects as well).

Put `GenerateProject.PlayStation4.CSharp.xslt` inside `%appdata%/.protobuild-xslt`.  When you generate for PlayStation4, it should use this file for generation.

Let me know if this works for you.

**Comment by tomspilman** [comment]
> Put `GenerateProject.PlayStation4.CSharp.xslt` inside `%appdata%/.protobuild-xslt`

You mean Windows OS user app data folder?

I can make due with this for now, but the XSLT has to exist in the MonoGame for PS4 repo in the future.  We can't have something outside of source control to install in some folder.  I don't see why this has to be any different than in the MonoGame public project which has custom XSLTs in it.

> Let me know if this works for you.

I'll get to this later this week and report back here.

**Comment by hach-que** [comment]
I would suggest a one-time setup Batch script to copy the file from the MonoGame PS4 repository into %appdata%\.protobuild-xslt (they're already going to be configuring other things manually for the project, so I don't think this is a huge concern).

**Comment by tomspilman** [comment]
> I would suggest a one-time setup Batch script 

I'd rather not.  Protobuild should really prefer a local XSLT when it exists over one installed on the machine.

**Comment by hach-que** [comment]
The local XSLTs are only used for that module (just as that version of Protobuild.exe is only used for that module).  We also can't just copy out the XSLT globally because if two modules both have `GenerateProject.PlayStation4.CSharp.xslt`, it's not clear which one should be used (which is why it'll be left up to the user).  We can't de-duplicate over module URLs either because in this scenario we're not playing the PlayStation 4 XSLT in it's own repository.

**Comment by tomspilman** [comment]
> The local XSLTs are only used for that module

If they would just apply to MonoGame that would be fine with me.

> if two modules both have  GenerateProject.PlayStation4.CSharp.xslt , it's not clear which one should be used

Throw an error or warning... that would be fine with me.

**Comment by hach-que** [comment]
Don't you need the game that's using MonoGame to also have the template applied to it?

**Comment by hach-que** [comment]
(Otherwise the game will just get generated as a Windows desktop project)

**Comment by tomspilman** [comment]
> Don't you need the game that's using MonoGame to also have the template applied to it? 

No... not at all.  Again all we care about is maintaining the MonoGame project.  Check out monogame, run Protobuild -generate Playstation4... open the solution and do work.

Everything else is already covered.

**Comment by hach-que** [comment]
Yeah I know that's all you care about, but I need to consider the broader context and use cases, and devise a solution that will work for that.

There's no point putting a hack in now for MonoGame, and then say in 12 months time when I'm developing for PS4, having to break your workflow in a newer release of Protobuild because it didn't consider the scenario of developing a PS4 game.

**Comment by tomspilman** [comment]
>  need to consider the broader context and use cases, and devise a solution that will work for that.

Sure... but we need to avoid breaking the workflow of myself and others that work with MonoGame on a daily basis for a feature none of use.

I want you to make Protobuild great and it can have all sorts of fantastic features.  I just need it to still do the simple work of a C# version of Premake without jumping thru a ton of extra hoops.

**Comment by hach-que** [comment]
This shouldn't be breaking the workflow that anyone currently uses, because none of this functionality exists. For PlayStation 4 you need to do a bunch of manual set-up and configuration of the project already; this reduces the amount of configuration and set-up significantly and only has one one-time setup (that you never need to do again) at the cost of ensuring that we don't (almost certainly) break a workflow in the future.

**Comment by tomspilman** [comment]
It is like I asked a mechanic to fix my flat tire and he keeps trying to sell me a new car. :)

**Comment by hach-que** [comment]
That's because she knows that even if she fixes the tire, the doors will fall off in 6 months and it's cheaper to replace it now instead of when you're driving down the highway at 100km/hr :)

**Comment by hach-que** [comment]
(Only in this case you can actually fix the tire yourself by modifying the Protobuild.exe that you distribute, but at the cost of knowing that future versions of Protobuild *will* break your workflow, which is why I'm so strongly advising against it)

**Comment by tomspilman** [comment]
@hach-que 

So I tried your branch...
  
  https://github.com/hach-que/Protobuild/tree/nda

First problem...  how to I add a new platform to it?  I tried this in the definition for the PS4 platform:

```
<?xml version="1.0" encoding="utf-8"?>
<IncludeProject Name="MonoGame.Framework.PS4" Path="MonoGame.Framework" AppliesTo="MonoGame.Framework" Platforms="PlayStation4">

  <ProjectGuids>
    <Platform Name="PlayStation4">09C41A48-7BF3-4A46-9EB8-CE95B4C27CA9</Platform>
  </ProjectGuids>

  <Properties>
	<CustomDefinitions>
      <Platform Name="PlayStation4">TRACE;PS4</Platform>	  
    </CustomDefinitions>
  </Properties>

  <Files>
    <Compile Include="Graphics\GraphicsDevice.PlayStation4.cs">
  </Files>

</Project>
```

**Comment by hach-que** [comment]
If that's in a submodule (or even just next to `MonoGame.Framework.definition` for testing purposes), and you generate for the main MonoGame repository, you should end up with `Graphics\GraphicsDevice.PlayStation4.cs` in the project.  That is, `Protobuild.exe --generate PlayStation4` should work.

I haven't implementing merging the properties in yet, so the `Properties` and `ProjectGuids` won't get read.  However, with the `IncludeProject` and putting an XSLT file in the `%appdata%\.protobuild-xslt`, you should be able to have something generate.

**Comment by tomspilman** [comment]
> If that's in a submodule 

It is not a git submodule if that is what you mean.  

I simply placed a folder within the MonoGame folder.  That folder includes a Protobuild.exe and a Build folder.  The Build folder has a module.xml that looks like this:

```
<?xml version="1.0" encoding="utf-8"?>
<Module>
  <Name>MonoGame.Framework</Name>
  <DefaultAction>Generate</DefaultAction>
  <DefaultWindowsPlatforms>PlayStation4</DefaultWindowsPlatforms>
  <SupportedPlatforms>PlayStation4</SupportedPlatforms>
  <DisableSynchronisation>true</DisableSynchronisation>
  <GenerateNuGetRepositories>false</GenerateNuGetRepositories>
</Module>
```

Then there is a Projects folder within that with the `MonoGame.Framework.definition` that I posted above.

When I run `Protobuild -generate Playstation4` it acts like it doesn't know Playstation4 is a platform.

Is Playstation4 hardcoded into Protobuild.exe or something?  Could I have a capitalization difference?


> I haven't implementing merging the properties in yet

Ok....  but that will be required in the end.

**Comment by hach-que** [comment]
I actually just ran into an issue I think with "AppliesTo" not applying to projects, so I'm investigating now.

**Comment by hach-que** [comment]
Oh of course I ran into that issue, I don't have the experimental version of Protobuild.

**Comment by hach-que** [comment]
Okay, so AppliesTo is working for me in the project creator.

You'll probably need to add PlayStation4 to the whitelist of supported platforms in MonoGame, otherwise you'll get the standard "this platform isn't supported error".

**Comment by hach-que** [comment]
FYI I've merged the `nda` branch into `master` because I need the `AppliesTo` feature in the project creator.  I'll be deleting the `nda` branch, and adding support for references and properties on the `master` branch for IncludeProjects instead.

**Comment by tomspilman** [comment]
> You'll probably need to add PlayStation4 to the whitelist of supported platforms in MonoGame

That is ok for now...  i'll try that and report back.  Eventually it would be good if even that was private.  This would allow us to more easily work on non-announced platforms.

> FYI I've merged the `nda` branch into `master`

Ok...  i'll work from there moving forward.

**Comment by tomspilman** [comment]
Ok... adding PS4 to the main MonoGame module under `<SupportedPlatforms>` helped.  It is now generating the project/solution.

But for some reason it is also generating the Windows projects/solutions too.  My command line is `Protobuild -generate Playstation4` ....   any idea why it is generating for Windows as well?

**Comment by tomspilman** [comment]
Yeah... this is all messed up.  The PS4 solution contains references to all the Windows versions of things too...  like MonoGame.Framework.Windows and MonoGame.Framework.Net.Windows.

So still i'm doing something really wrong.

How the hell do you figure out what is causing it to drag in these other projects?

**Comment by hach-que** [comment]
That is expected as of the latest version of Protobuild. We include the
host platform's projects in all solutions that don't target the host
platform.

**Comment by hach-que** [comment]
(it doesn't reference the Windows projects though, they're just included in
the solution. PS4 projects will still only reference other PS4 projects)

**Comment by tomspilman** [comment]
> We include the host platform's projects in all solutions that don't target the host
platform.

Can you add a way to turn that off?  I don't need 5 or 6 projects to a platform i'm not targeting open in the PS4 solution.

**Comment by tomspilman** [comment]
Starting to get some results...  but still ugly stuff happening.

Like currently it generates two solutions... one the main MonoGame folder called `MonoGame.Framework.PS4.sln` and another in the PS4 submodule called `MonoGame.PS4.PS4.sln` which is empty referencing no projects.

How to I tell it not to generate that extra empty solution?

**Comment by tomspilman** [comment]
So I am starting to get some reasonable results with this.  I have a subfolder within the `MonoGame` folder which is called 'PS4' and has the content of the PS4 repo.  I was able to include some PS4 specific C# files from this other folder as links which was the first goal.

Still need to see if external project references work.

What is missing/sucks so far:

  - No support for `<ProjectGuids>` or `<Properties>` within the `<IncludeProject>`. (high priority)
  - You cannot add a `Platform="PS4"` to the `<IncludeProject>` to make sure it is completely excluded for non-PS4 platforms. (med priority)
  - It generates unnecessary Windows projects in the PS4 solution. (med priority)
  - Have to add PS4 to the `SupportedPlatforms` in the main MonoGame repo (low priority).

So I finally have some hope this will work out.  Thanks for the help @hach-que !

**Comment by hach-que** [comment]
> You cannot add a Platform="PS4" to the <IncludeProject> to make sure it is completely excluded for non-PS4 platforms. (med priority)

This is now done in upstream.

> Nevermind... figured it out. It was because I had a Protobuild.exe in the submodule which I didn't need.
> It generates unnecessary Windows projects in the PS4 solution. (med priority)

So you should have just one Protobuild.exe in the submodule (in the root of the repo), alongside the Build folder.  Inside that build folder there should be Module.xml, and the Projects folder.  Inside that Projects folder, there should be just an include project.  If the PS4 submodule doesn't have Protobuild inside it, then the include project won't be read.

If you're seeing it generate any kind of solution in that submodule in this scenario, then that is a bug that needs fixing.  What you should see while generating MonoGame is something like:

```
...
Skipping submodule generation for MonoGame.PS4 (there are no projects to generate)
...
```

Can you paste the output of running --generate in the MonoGame repository (with MonoGame.PS4 as a submodule)?

**Comment by hach-que** [comment]
> Have to add PS4 to the SupportedPlatforms in the main MonoGame repo (low priority).

Not sure we'll end up with a way around this.  I don't really want to do a recursive scan of submodules just to check if the platform is valid (we already traverse recursively during generation, so adding another check here would approximately double the IO operations searching for submodules).

**Comment by tomspilman** [comment]
> If the PS4 submodule doesn't have Protobuild inside it, then the include project won't be read.

That isn't what i'm seeing.  I just have one Protobuild.exe in the MonoGame folder.  The child PS4 folder has only a Build folder and no Protobuild.exe.


> Can you paste the output of running --generate in the MonoGame repository 

```
T:\Projects\MonoGame>Protobuild.exe -generate PS4
Starting resolution of packages...
Invoking package resolution in submodule for MonoGame.PS4
Finished submodule package resolution for MonoGame.PS4
Package resolution complete.
Starting generation of projects for PS4
Loading: 2MGFX
Loading: 2MGFXReferences
Loading: Framework.Content.Pipeline.References
Loading: FrameworkReferences
Loading: FrameworkReferences.Net
Loading: Lidgren.Network
Loading: Lidgren.Network.References
Loading: MGCB
Loading: MonoGame.Framework.Content.Pipeline
Loading: MonoGame.Framework
Loading: MonoGame.Framework.Net
Loading: MonoGame.Tests
Loading: MonoGame.Tests.References
Loading: Pipeline
Loading: PipelineReferences
Loading: MonoGame.Framework.PS4
Invoking submodule generation for MonoGame.PS4
Finished submodule generation for MonoGame.PS4
Generating: MonoGame.Framework
NOTE: Added MonoGame.Framework.PS4 as a reference in MonoGame.Framework because of an AppliesTo attribute on the include
 project.
Generating: (solution)
Generation complete.
Starting resolution of packages...
Invoking package resolution in submodule for MonoGame.PS4
Finished submodule package resolution for MonoGame.PS4
Package resolution complete.
Starting generation of projects for Windows
Loading: 2MGFX
Loading: 2MGFXReferences
Loading: Framework.Content.Pipeline.References
Loading: FrameworkReferences
Loading: FrameworkReferences.Net
Loading: Lidgren.Network
Loading: Lidgren.Network.References
Loading: MGCB
Loading: MonoGame.Framework.Content.Pipeline
Loading: MonoGame.Framework
Loading: MonoGame.Framework.Net
Loading: MonoGame.Tests
Loading: MonoGame.Tests.References
Loading: Pipeline
Loading: PipelineReferences
Loading: MonoGame.Framework.PS4
Enabled service: MonoGame.Framework/DirectXGraphics
Enabled service: MonoGame.Framework/XAudioAudio
Enabled service: MonoGame.Framework/_XNADesignProvided
Enabled service: MonoGame.Framework/_FrameworkDispatcherProvided
Invoking submodule generation for MonoGame.PS4
Finished submodule generation for MonoGame.PS4
Generating: 2MGFX
Generating: Lidgren.Network
Generating: MGCB
Generating: MonoGame.Framework.Content.Pipeline
Generating: MonoGame.Framework
NOTE: Added MonoGame.Framework.PS4 as a reference in MonoGame.Framework because of an AppliesTo attribute on the include
 project.
Generating: MonoGame.Framework.Net
Generating: MonoGame.Tests
Generating: Pipeline
Generating: (solution)
Generation complete.
```

**Comment by tomspilman** [comment]
> I don't really want to do a recursive scan of submodules just to check if the platform is valid

Right...  that does suck.

How about instead you assume the platform is correct... proceeding to read modules.... until you determine the platform doesn't exist.   It just means it takes a little longer to fail...  if you got the platform right it is just as fast as it has been.

**Comment by hach-que** [comment]
I'm assuming the version of Protobuild in the MonoGame folder is the
latest?

**Comment by tomspilman** [comment]
I updated last from the nda branch version.  I'll try the latest now.

**Comment by tomspilman** [comment]
I see no difference with the latest Protobuild.exe:

```
T:\Projects\MonoGame>Protobuild.exe -generate PS4
Starting resolution of packages for PS4...
Invoking package resolution in submodule for MonoGame.PS4
Finished submodule package resolution for MonoGame.PS4
Package resolution complete.
Starting generation of projects for PS4
Loading: 2MGFX
Loading: 2MGFXReferences
Loading: Framework.Content.Pipeline.References
Loading: FrameworkReferences
Loading: FrameworkReferences.Net
Loading: Lidgren.Network
Loading: Lidgren.Network.References
Loading: MGCB
Loading: MonoGame.Framework.Content.Pipeline
Loading: MonoGame.Framework
Loading: MonoGame.Framework.Net
Loading: MonoGame.Tests
Loading: MonoGame.Tests.References
Loading: Pipeline
Loading: PipelineReferences
Loading: MonoGame.Framework.PS4
Invoking submodule generation for MonoGame.PS4
Finished submodule generation for MonoGame.PS4
Generating: MonoGame.Framework
NOTE: Added MonoGame.Framework.PS4 as a reference in MonoGame.Framework because of an AppliesTo attribute on the include
 project.
Generating: (solution)
Generation complete.
Starting resolution of packages for Windows...
Invoking package resolution in submodule for MonoGame.PS4
Finished submodule package resolution for MonoGame.PS4
Package resolution complete.
Starting generation of projects for Windows
Loading: 2MGFX
Loading: 2MGFXReferences
Loading: Framework.Content.Pipeline.References
Loading: FrameworkReferences
Loading: FrameworkReferences.Net
Loading: Lidgren.Network
Loading: Lidgren.Network.References
Loading: MGCB
Loading: MonoGame.Framework.Content.Pipeline
Loading: MonoGame.Framework
Loading: MonoGame.Framework.Net
Loading: MonoGame.Tests
Loading: MonoGame.Tests.References
Loading: Pipeline
Loading: PipelineReferences
Loading: MonoGame.Framework.PS4
Enabled service: MonoGame.Framework/DirectXGraphics
Enabled service: MonoGame.Framework/XAudioAudio
Enabled service: MonoGame.Framework/_XNADesignProvided
Enabled service: MonoGame.Framework/_FrameworkDispatcherProvided
Invoking submodule generation for MonoGame.PS4
Finished submodule generation for MonoGame.PS4
Generating: 2MGFX
Generating: Lidgren.Network
Generating: MGCB
Generating: MonoGame.Framework.Content.Pipeline
Generating: MonoGame.Framework
NOTE: Added MonoGame.Framework.PS4 as a reference in MonoGame.Framework because of an AppliesTo attribute on the include
 project.
Generating: MonoGame.Framework.Net
Generating: MonoGame.Tests
Generating: Pipeline
Generating: (solution)
Generation complete.
```

**Comment by tomspilman** [comment]
Note the `Platform="PS4"` bit doesn't seem to work.  If definition looks like this:

```
<?xml version="1.0" encoding="utf-8"?>
<IncludeProject Name="MonoGame.Framework.PS4" Path="MonoGame.Framework" AppliesTo="MonoGame.Framework" Platform="PS4">

  <ProjectGuids>
    <Platform Name="PS4">5B8FCDD7-3A9E-47B2-B68A-D9B4B49D62F7</Platform>
  </ProjectGuids>

  <Properties>
    <CustomDefinitions>
      <Platform Name="PS4">TRACE;PS4</Platform>	  
    </CustomDefinitions>
  </Properties>

  <Files>

    <Compile Include="Graphics\GraphicsDevice.PS4.cs">
      <Link>Graphics\GraphicsDevice.PS4.cs</Link>
    </Compile>

  </Files>
</IncludeProject>
```
It still includes `GraphicsDevice.PS4.cs` into the Windows csproj too.

**Comment by hach-que** [comment]
> How about instead you assume the platform is correct... proceeding to read modules.... until you determine the platform doesn't exist. It just means it takes a little longer to fail... if you got the platform right it is just as fast as it has been.

That's the default behaviour if you omit the `<SupportedPlatforms>` tag from a module.

**Comment by tomspilman** [comment]
> That's the default behaviour if you omit the `<SupportedPlatforms>` tag from a module.

Interesting...  i'll experiment with that.  Again that is sort of low priority.

At some point we will have platforms we're developing for that we cannot mention publicly.  The ability to not include *any* reference to that platform from the main MonoGame repo is a desirable feature.

**Comment by tomspilman** [comment]
Another issue that is becoming more apparent is we need to setup the XSLTs to use features rather than platforms to decide how to generate the project.  We can then put some other configuration settings in possibly the Module.xml if not the actual .definition that controls how projects/solutions are generated for each platform.

This again is to make it easier to add platforms without a whole XSLT needing to be maintained.

**Comment by hach-que** [comment]
I don't quite understand the request - you need the XSLT to do the generation.

There's `AdditionalProjectTransforms.xslt` which is a file that gets inserted into the XSLT.  This is usually used for additional build configuration for certain platforms under certain projects; it is documented here: https://protobuild.readthedocs.org/en/latest/customizing_protobuild.html#specifying-additional-transforms

**Comment by tomspilman** [comment]
> There's  AdditionalProjectTransforms.xslt 

That might work.

The issue is all about platform hiding...  we can't put into the public repo NDA platform information.  So being able to inject that stuff from the include/applyto project is what I am needing.

**Comment by hach-que** [comment]
For the PS4 XSLT you should be placing it in `%appdata%\.protobuild-xslt\GenerateProject.CSharp.PlayStation4.xslt`.  Then when you generate for PlayStation4 it will use that XSLT template, regardless of where or which Protobuild project you are generating.

**Comment by tomspilman** [comment]
> For the PS4 XSLT you should be placing it in `%appdata%`

We can't do that.  It has to be in the repo.

**Comment by hach-que** [comment]
You can't place it in the repo; the repo is public (and the generation needs to work across submodules).

**Comment by tomspilman** [comment]
The PS4 repo is not public.  We need to be able to place it there.

**Comment by hach-que** [comment]
We can't place it there, it will not work for other dependencies the game has.

**Comment by tomspilman** [comment]
That isn't a concern...  the XSLT is exclusively for the PS4 MonoGame project.  Not for any other game or library.

**Comment by hach-que** [comment]
Games being made with MonoGame for PS4 need to be able to use it?

**Comment by tomspilman** [comment]
They don't.

**Comment by hach-que** [comment]
How will they have the correct project format and reference the assembly?  If you're just shipping binaries  (for MonoGame PS4) you can just configure your build server to copy the XSLT into the public repo's Build folder before you run the build?

**Comment by tomspilman** [comment]
> How will they have the correct project format and reference the assembly?

We already have a solution for that.  All I need is for the MonoGame PS4 project to be able to generate its XSLTs.  Nothing more.

**Comment by hach-que** [comment]
Okay, in that case just configure your build server to copy the XSLT into the public repo's Build folder after it's been cloned (and for debugging just copy it in on your local computer, just don't commit it).

**Comment by tomspilman** [comment]
> Okay, in that case just configure your build server to copy the XSLT 

This has nothing to do with the build server.

We need the XSLTs to be in the PS4 repo and for Protobuild to pick them up.  I don't need a solution that copies them anywhere on my PC.

I need this feature. This isn't an impossible task.

**Comment by tomspilman** [comment]
Here is what could happen maybe.  When Protobuild scans the modules it identifies these addon XSLTs and Protobuild itself installs them to `%appdata%`.  The only issue there is Protobuild being smart enough to update them when the XSLT changes or dealing with conflicts.

**Comment by tomspilman** [comment]
This really all sidesteps the issues.  The XSLTs are a brute force tool.  Yes you can do anything with them, but often there are few things you actually need to do...  set compiler options, add properties, set the VS version, set the tool version, etc.

Worse when Protobuild adds a new feature we need to figure out how to merge the custom XSLTs with the local changes...  that is when we remember to update them.  We often end up for weeks of months with a broken feature that is XSLT driven.

**Comment by hach-que** [comment]
> Here is what could happen maybe. When Protobuild scans the modules it identifies these addon XSLTs and Protobuild itself installs them to %appdata%. The only issue there is Protobuild being smart enough to update them when the XSLT changes or dealing with conflicts.

Yes, the issue is conflicts, and dealing with how to resolve them.

Protobuild already has a mechanism for dealing with such conflicts, and as much as you don't like the idea/concept/etc of it, it is the officially supported way of de-duplicating multiple references to the same artifact (whether that artifact is a third-party library, a Git repository, etc.)  This mechanism already works and is well tested.  As I have suggested previously, you could put the XSLT in a seperate Git repository by itself, and add a reference to it (in the `MonoGame.PS4` private repo) like this:

```
  <Packages>
    <!-- .... -->
    <Package Uri="https-git://github.com/MonoGame/PS4XSLT.git" Folder="PS4XSLT" GitRef="master" Type="PlatformExtension" />
    <!-- .... -->
  </Packages>
```

Protobuild can then de-duplicate the template for PS4 correctly, across all the projects that are using MonoGame's version of the PS4 XSLT.

As I've stated a hundred times, this is functionally the same thing as doing `git submodule add https://github.com/MonoGame/PS4XSLT.git`; the _only_ difference is that Protobuild is aware of it, can de-duplicate any references that are identical in the hierarchy and can pull in XSLT files without traversing fully up and down the file-system hierarchy.

> This really all sidesteps the issues. The XSLTs are a brute force tool. Yes you can do anything with them, but often there are few things you actually need to do... set compiler options, add properties, set the VS version, set the tool version, etc.
> 
> Worse when Protobuild adds a new feature we need to figure out how to merge the custom XSLTs with the local changes... that is when we remember to update them. We often end up for weeks of months with a broken feature that is XSLT driven.

There is no other viable way to do this.  There are potentially hundreds of points where you would need to support "additional XSLT"; off the top of my head:

* ToolsVersion
* ProjectVersion
* Platform/Configuration sections (and in the solution as well)
* Additional references that the platform implies
* Additional files that are required to be included to compile on a platform
* Setting of arbitrary and various properties that are set in various different sections of the file (like architecture, debug modes, etc.) where the available properties change per-platform
* Post-build hooks, which are different per-platform
* Just plain weird stuff like the stripping of XNA references for Windows Phone
* NuGet references

To support any additional platform, we'd need to have injection options at all these points and more.  Maintaining that would be a nightmare, not to mention if you needed to track down where a particular value is coming from.  This also amplifies the "resolving conflicts" issue a hundred-fold because now you need to resolve conflicts in hundreds of places, rather than just resolving the conflict of "path to the right XSLT file".

**Comment by tomspilman** [comment]
> As I have suggested previously, you could put the XSLT in a seperate Git repository 

Could Protobuild then not pull the XSLT from the existing MonoGame.PS4 repo?

This is as long as all it uses this for is pulling the XSLT.  I don't need Protobuild managing my Git repositories for any reason.

**Comment by hach-que** [comment]
Protobuild **isn't** managing your Git repositories.  It's using the repository URL as an identifier to de-duplicate the content of the repository, so that when multiple projects reference the same URL, it knows they're the same thing and that there isn't a conflict.

**Comment by tomspilman** [comment]
> Protobuild isn't managing your Git repositories.

So it isn't going to clone that repo and place a copy on my PC?

**Comment by hach-que** [comment]
Not if it's already there (or has already been cloned by Git).

**Comment by tomspilman** [comment]
So it will try to clone and manage repos if I don't beat it to the punch.

I just need a way for the XSLT for PS4 and other platforms I cannot talk about to exist in their respective repos.  I don't need additional repos for XSLTs.

Can we get...
````
<Package Uri="https-git://github.com/MonoGame/PS4XSLT.git" Folder="PS4XSLT" GitRef="master" Type="PlatformExtension" />
```
To pull a single file or folder from a repo instead of the whole repo?

Is `PlatformExtension` a code word for platform XSLT and it knows where to place it?

**Comment by hach-que** [comment]
> So it will try to clone and manage repos if I don't beat it to the punch.

I don't understand what you mean by this - the XSLT is in that repository, what do you expect to happen if the folder that the XSLT is in hasn't already been cloned by Git? (i.e. the user hasn't initialized Git submodules)  The file literally doesn't exist on your computer otherwise, so what is it meant to do?

> Can we get...
>
> ```
> <Package Uri="https-git://github.com/MonoGame/PS4XSLT.git" Folder="PS4XSLT" GitRef="master" Type="PlatformExtension" />
> ```
>
> To pull a single file or folder from a repo instead of the whole repo?

We could add a package type that pulls a single file from a web server (you could then just make this a URL that points to the raw link on GitHub, e.g. like https://github.com/hach-que/Protobuild/raw/master/Protobuild.exe).  This would probably end up being:

```
<Package Uri="https-file://github.com/MonoGame/MonoGame.PS4/XSLT/GenerateProject.CSharp.PlayStation4.xslt" Folder="PS4XSLT" GitRef="master" Type="PlatformExtension" />
```

> Is `PlatformExtension` a code word for platform XSLT and it knows where to place it?

Yeah the intention here is that Protobuild would know that the target repo (or file) just contains an XSLT and can treat it differently to other repos (where it would be expecting them to be a Protobuild submodule that it needs to generation in).  Roughly it could see that in the hierarchy and go, okay, now I have a reference that will provide me with an additional XSLT, so I will copy it to whatever location I need (whether that's passing state through to other submodules, storing it locally, copying, etc. doesn't matter as long as it's providing that reference to all the modules in the hierarchy without them having to perform disk scans).

**Comment by tomspilman** [comment]
> the XSLT is in that repository, what do you expect to happen if the folder that the XSLT is in hasn't already been cloned by Git?

I confused the `<Package>` in this context with the thing in here:

https://github.com/mono/MonoGame/blob/develop/Build/Module.xml#L12

... which I really dislike.

It is the same story with me... whenever Protobuild steps beyond project generation I will be unhappy.

> We could add a package type that pulls a single file from a web server

That sounds like it could work...  but a final question.  Where would that `<Package>` line go?  It has to be in the module in the PS4 repo.  It cannot be in the public MonoGame repo.

**Comment by hach-que** [comment]
> It is the same story with me... whenever Protobuild steps beyond project generation I will be unhappy.

A part of project generation is knowing what dependencies you have - in this case we depend on an external XSLT file.  The `<Package>` reference is the way of specifying things you depend on, regardless of what they are.

> That sounds like it could work... but a final question. Where would that <Package> line go? It has to be in the module in the PS4 repo. It cannot be in the public MonoGame repo.

Yes it would be in the private PS4 repo.

**Comment by tomspilman** [comment]
> The `<Package>` reference is the way of specifying things you depend on, regardless of what they are.

What we disagree on here is if Protobuild should be doing git operations.  IMO it should not.  It limits Protobuild to being a tool that only works with Git.  What if a project uses Mercurial?  Or SVN?  Or is uses Perforce?  Or something else we've never heard of?

Project generation should have nothing to do with what source control system you choose to support.

> Yes it would be in the private PS4 repo.

Perfect...  then it sounds like maybe we have a viable solution at long last.

**Comment by tomspilman** [comment]
Ok here is a tricky case.  In the public MonoGame repo we have this defined:

https://github.com/mono/MonoGame/blob/develop/Build/Projects/MonoGame.Framework.definition#L509
```
<Compile Include="Graphics\RenderTarget3D.cs"> 
   <Services>DirectXGraphics</Services> 
</Compile> 
```
How would an include/applyto project add itself to the platform list for that file?

I tried simply adding this to the include project:
```
    <Compile Include="Graphics\RenderTarget3D.cs" />
```
But it seems Protobuild skips including it.  Also tried this:
```
    <Compile Include="Graphics\RenderTarget3D.cs">
      <Platforms>PS4</Platforms>
    </Compile>
```
Maybe because it isn't combining the platform info correctly with the public .definition in the MonoGame repo?

**Comment by hach-que** [comment]
> What we disagree on here is if Protobuild should be doing git operations. IMO it should not. It limits Protobuild to being a tool that only works with Git. What if a project uses Mercurial? Or SVN? Or is uses Perforce? Or something else we've never heard of?

Protobuild's package management isn't limited to Git; we could quite easily support `https-hg://` or `https-svn://`.  But no-one has expressed an interested in this.  That's also only necessary if the target project you're consuming is in one of those source control types, you can use this functionality from your project regardless of what source control system it's using.

Of course, you can also reference a package with `https-nuget://` and be source control independent, because that only relies on HTTP requests.  Or you can reference a binary package with `https://` and be source control independent, because that also replies on HTTP requests.

I'm open for a way of making this less dependent on specific systems, but outside "support other source control systems", I don't know how to address this.

> Ok here is a tricky case. In the public MonoGame repo we have this defined:
> 
> https://github.com/mono/MonoGame/blob/develop/Build/Projects/MonoGame.Framework.definition#L509
> 
> ```
> <Compile Include="Graphics\RenderTarget3D.cs"> 
>    <Services>DirectXGraphics</Services> 
> </Compile> 
> ```
> 
> How would an include/applyto project add itself to the platform list for that file?
> 
> I tried simply adding this to the include project:
> 
> ```
>     <Compile Include="Graphics\RenderTarget3D.cs" />
> ```
> But it seems Protobuild skips including it. Also tried this:
> 
> ```
>     <Compile Include="Graphics\RenderTarget3D.cs">
>       <Platforms>PS4</Platforms>
>     </Compile>
> ```
> 
> Maybe because it isn't combining the platform info correctly with the public .definition in the MonoGame repo?

Hmm, that is a tricky situation to resolve.  The first one is getting turned off because of the `<Services>` attribute, that doesn't explain why you don't get a second addition from the include project though - Protobuild doesn't de-duplicate any file entries here...

I'm guessing when you Ctrl-F, the RenderTarget3D doesn't appear in the generated PS4 C# project at all right?

**Comment by tomspilman** [comment]
> I'm guessing when you Ctrl-F, the RenderTarget3D doesn't appear in the generated PS4 C# project at all right?

Right...  it doesn't show up in the project in VS.  However!

If I open the project in notepad I find this:

```
    <Compile Include="..\PS4\MonoGame.Framework\Graphics\RenderTarget3D.cs">
      <Link>Included Code\PS4/MonoGame.Framework\Graphics\RenderTarget3D.cs</Link>
      <Platforms>PS4</Platforms>
      <FromIncludeProject>True</FromIncludeProject>
    </Compile>
```
So it seems like Protobuild assumed it was a linked file and generated this weird link and not just a regular reference to the source.

**Comment by hach-que** [comment]
The link is intentional because it's referencing a file outside the project
folder; while we're at it we nest it under an "Included Code\<project
name>" so it's easier to see where included files are coming from. However,
I did notice this being a bit weird when I was working on the project
creator stuff.

Can you try adding an empty `<Link />` tag underneath the Compile node in
the include project and see if it appears in VS then?

**Comment by DDReaper** [comment]
On my two pence thought. If it does link to external repos, then it could be hosted on a private repo with restricted access, which solves any security access issue.

**Comment by hach-que** [comment]
Yeah that's basically the plan; we'll use the security of the repo to
prevent unauthorized access

**Comment by tomspilman** [comment]
> Can you try adding an empty `<Link />` tag 

Ok this worked...
```
    <Compile Include="..\..\MonoGame.Framework\Graphics\RenderTarget3D.cs">
      <Link/>
    </Compile>
```
Ideally we could fix the "Included Code" thing and the `<Link/>` wouldn't be necessary here.

**Comment by hach-que** [comment]
Oh yeah, the path for the include project items is relative to the include path, not the project that's having it included.  So having to do `..\..\` is totally expected there.  The `<Link/>` stuff is a bug though.

**Comment by tomspilman** [comment]
So my major remaining issues with Protobuild here are:

  - No support for `<ProjectGuids>` or `<Properties>` within the `<IncludeProject>`.
  - It generates unnecessary Windows projects in the PS4 solution.
  - The `<Package>` fix for pulling an XSLTs from an existing repo.
  - Fix the `<Link/>` required when doing relative paths to source files.

**Comment by hach-que** [comment]
My plan is to return to this in the near future, but for the moment I'm focused on getting #4295 finished and merged first.

**Comment by tomspilman** [comment]
> I'm focused on getting #4295 finished and merged first.

It would help console development tremendously to get this done first.  I can't say when we'll be ready to merge #4295, but these features were needed months ago.

**Comment by hach-que** [comment]
Yes I know, but I can't put any more things on my plate at the moment because I'm so overloaded with project work, and since I'll be going back to full-time work on January 6th it puts even more pressure on things.  I need to get things done and cleared off my to do list before I start trying to do more things in parallel.

**Comment by tomspilman** [comment]
I guess I will have to maintain a hacked together system for longer while we wait for Protobuild.

**Comment by tomspilman** [comment]
@hach-que - Another issue that came up....  I need to be able to extend the project references.  Like the stuff we include in...

https://github.com/mono/MonoGame/blob/develop/Build/Projects/FrameworkReferences.definition

Simply being able to add a  `<References>` section within the IncludeProject would work fine.

**Comment by hach-que** [comment]
Yeah this is on my list, but I start full-time work again tomorrow so my time will be limited.

**Comment by tomspilman** [comment]
> I start full-time work again tomorrow so my time will be limited.

I understand that all too well.  I just urge you to prioritize these issues over another other new MonoGame feature.  It reflects very poorly to the console manufacturers that we have to manually hack at MonoGame/Protobuild to make things work.

**Comment by hach-que** [comment]
Sony have been completely unresponsive in terms of applying for a license, which is not helping the situation at all. I applied for SDK access back at the end of November, the auto-reply said 5 working days and I never heard anything since. 

While having access to the SDK isn't strictly required, it will make things go faster because I can just do the work rather than lots of back and forth with people who do have SDK access, and it means I can actually experience the issues that are being described so I can better address them.

**Comment by hach-que** [comment]
To clarify, the main blocker is getting my other MonoGame work merged at the moment because I need to get things off my plate. That is blocked by 3.5 being released before those PRs can be merged.  Getting SDK access now won't help the situation, I'm just bringing it up because it's been frustratingly difficult to get access to, and it might have helped things a few months ago.

**Comment by tomspilman** [comment]
> Sony have been completely unresponsive in terms of applying for a license

I can try to help you with that... but these features have nothing to do with Sony.  Read that again... it has nothing to do with Sony.  I can't say more than that.

> To clarify, the main blocker is getting my other MonoGame work merged at the moment because I need to get things off my plate. 

Sure...   and there is some effort at that.  But #4295 is not higher priority.  If you are planning to hold out work on this until #4295 I merged let me know now.  I will have to find another solution for consoles than Protobuild.

Remember I brought up these issues almost a year ago now.  By Protobuild not adding these features it is hurting MonoGame on consoles.

**Comment by hach-que** [comment]
I know and I'm trying to make time for this stuff. My intention was for #4295 to be ready and merged by mid-December.

But it also took way way long than I intended (like 2-3x longer), and there was some competitive pressure to get it all fully implemented now rather than just a first version, which pushed both this AND a lot of other unrelated non-MonoGame work back.

I want to emphasize that #4295 had far more wide-reaching impacts on my schedule than just this and now I just want to be done with it.  A lot of commercial development stuff happening in my company suffered for spending time on it.

Which part of this is the most pressing for you?  I'll see if I can spend some time on it this weekend. I'm guessing the References + properties support since that will remove any need to patch the public repo other than copying an XSLT file across (which is not ideal but is still leagues better than patching files).

**Comment by tomspilman** [comment]
> I want to emphasize that #4295 had far more wide-reaching impacts on my schedule than just this and now I just want to be done with it. 

I understand...  there are certainly optional projects I've taken on that have consume much more time and effort than expected.  We just have to hope the effort and problems we create for ourselves in the short term pays off in the end.

You'll know what I mean later once we make things public. ;)

> Which part of this is the most pressing for you? I'll see if I can spend some time on it this weekend. I'm guessing the References + properties support since that will remove any need to patch the public repo other than copying an XSLT file across 

That would be great.

Yes support for `<ProjectGuids>`, `<Properties>`, `<References>` within the include/addto project is most pressing stuff.

**Comment by hach-que** [comment]
Maybe this was previously mentioned, what's the need for ProjectGuids in this context? Protobuild will generate them for you if they aren't specified.

I only ask because I *think* that might be more complicated than the other two because of the way project GUIDs interact across modules, and if we can get away without doing it that'd be great.

**Comment by tomspilman** [comment]
> what's the need for ProjectGuids in this context?

I guess then it isn't required for now then.  However i would like to see that working in the future as we want the GUIDs to be consistent regardless to changes to Protobuild.

**Comment by hach-que** [comment]
We'll never ever ever **ever** change the project GUID generation in Protobuild. If I can guarantee you one thing that will never change, it is this.

We are fundamentally prevented from changing it because any change to that logic will break project generation across submodules with different versions of Protobuild.  This is because the parent Protobuild must be able to calculate the project GUID when generating the top-level solution for C# projects generating by the Protobuild that's in a submodule.

**Comment by hach-que** [comment]
To emphasize how sensitive the project GUID generation is (and how much we can't change it), it's one of the few things that gets singled out in the contributor documentation: https://protobuild.readthedocs.org/en/latest/contributing.html

It is part of the guaranteed backwards and forwards compatibility between modules that Protobuild provides.

**Comment by tomspilman** [comment]
> If I can guarantee you one thing that will never change, it is this.

That makes me feel better about it.... i'll look to not depend on it on these new platforms.

**Comment by hach-que** [comment]
I got some time this morning, I have not tested this *at all*, but here's an experimental version that should support `<References>` and `<Properties>` sections in include projects: https://github.com/hach-que/Protobuild/blob/nda-props/Protobuild.exe

Before it gets merged into master I want to make sure it works and add some functional tests around it for edge cases (like include projects referencing external projects which in turn reference other external projects), but let me know if there's anything you need changed before then.

**Comment by tomspilman** [comment]
> here's an experimental version 

Great...  I'll give it a try and report back a little later today.

**Comment by tomspilman** [comment]
@hach-que - Looks to be working great so far.  I'll dig in deeper and have more to say by tomorrow.

**Comment by hach-que** [comment]
Properties and references functionality is now merged into `master` (so you can pull that version into the public MonoGame repo now).

**Comment by tomspilman** [comment]
> Properties and references functionality for include projects is now merged into  master  

Great...  gonna wait a few days first and make sure I don't run into any other bugs with the console stuff.  Just want to cut down on unnecessary EXE updates.

**Comment by tomspilman** [comment]
@hach-que - So things have been working great so far.

The IncludeProject/AppliesTo stuff is working ok as is with the references and properties support.

The extra empty `<Link/>` needed at times to include some files is annoying, but works for now.

All this is letting me fully extend the framework and even tools without changes to the main repo.  A fantastic feature for development of new secret/private platforms.

The only issue that still requires me to patch the public repo is adding the new platform to `module.xml` `<SupportedPlatforms>` property and the `MonoGame.Framework.definition` `Platforms` attribute.  Being able to extend that somehow from a direct sub folder/module of the MonoGame repo would complete the extension system.

**Comment by hach-que** [comment]
That's good to hear.

I'll have to address any remaining issues after the Global Game Jam runs at the end of January.

**Comment by tomspilman** [comment]
@hach-que - Any time to revisit this?

The big remaining issue is the current requirement of adding the new platform names to `module.xml`'s `<SupportedPlatforms>` property and the `MonoGame.Framework.definition` `Platforms` attribute.  We need a way to extend those with a new platform, but not modify the MonoGame `module.xml` or `MonoGame.Framework.definition`.

**Comment by hach-que** [comment]
We're currently in full blown pre-production for a game.  I might get time to visit this stuff if I have to touch Protobuild, but otherwise it might not happen for a while because we are very time constrained.

I'm open to PRs that will solve these issues though - I just don't have time (at the moment) to write the code myself.  The former should be addressed by scanning submodules for maybe some `<AddsSupportedPlatforms>` attribute in the submodule's `Module.xml` file which then applies up the chain, and the latter by merging the `Platforms` attribute on applied include projects into the project the include project is applied to.

**Comment by tomspilman** [comment]
> I'm open to PRs that will solve these issues though

 No problem... we'll see what we can do.


> The former should be addressed by scanning submodules for maybe some `<AddsSupportedPlatforms>` attribute in the submodule's `Module.xml` file which then applies up the chain

That sounds correct to me.

> the latter by merging the `Platforms` attribute on applied include projects into the project the include project is applied to.

Yeah...  that should solve that nicely.

**Comment by tomspilman** [comment]
This worked pretty well as we left it.  We won't be changing it any further. https://github.com/MonoGame/MonoGame/issues/6207
